### ISSUE TITLE
AsyncStream & AsyncThrowingStream behavioral divergences & misleading documentation

### ISSUE THREAD

**Issue by jamieQ** [swiftlang/swift#76547]
### Description

issues:

1. AsyncStream supports multiple concurrent consumption, but its documentation states that it does not
2. AsyncThrowingStream does not support multiple concurrent consumption (its documentation is accurate)

### Reproduction

```swift
// non-throwing stream
func test_stream() async {
  let (s, c) = AsyncStream<Int>.makeStream()

  let t1 = Task { @MainActor in
    for await _ in s {}
  }
  await MainActor.run {}
  print("installed consumer 1")

  let t2 = Task { @MainActor in
    for await _ in s {}
  }
  await MainActor.run {}
  print("installed consumer 2")
  // works
}

// throwing stream
func test_throwing_stream() async {
  let (s, c) = AsyncThrowingStream<Int, Error>.makeStream()

  let t1 = Task { @MainActor in
    for try await _ in s {}
  }
  await MainActor.run {}
  print("installed consumer 1")

  let t2 = Task { @MainActor in
    for try await _ in s {}
  }
  await MainActor.run {}
  print("installed consumer 2")
  // fatal error shortly after this point
}

Task { await test_stream() }
Task { await test_throwing_stream() }
```

### Expected behavior

the async stream implementations should function analogously with respect to multi-consumption, and the corresponding documentation should accurately reflect their behaviors. in particular, the throwing variant should behave in the same manner that the non-throwing stream does.

### Environment

Apple Swift version 5.10 (swiftlang-5.10.0.13 clang-1500.3.9.4)
Target: arm64-apple-macosx14.0

### Additional information

relevant forum posts:

1. https://forums.swift.org/t/consuming-an-asyncstream-from-multiple-tasks-redux/69818
2. https://forums.swift.org/t/question-how-to-re-introduce-concurrent-iteration-runtime-error-for-asyncstream/71831

**Comment by malhal** [comment]
I believe the hypothesis that "AsyncStream supports multiple concurrent consumption" is incorrect, e.g. in the example below only a single consumer receives:
```
func test_stream() async {
    let (s, c) = AsyncStream<Int>.makeStream()
    
    let t1 = Task { @MainActor in
        for await x in s {
            
            print("consumer 1 received \(x)")
        }
    }
    await MainActor.run {}
    print("installed consumer 1")
    
    let t2 = Task { @MainActor in
        for await x in s {
            print("consumer 2 received \(x)")
        }
    }
    await MainActor.run {}
    print("installed consumer 2")
    
    c.yield(1) // only consumer 1 receives
}
```
Output
```
installed consumer 1
installed consumer 2
consumer 1 received 1
```

**Comment by jamieQ** [comment]
> I believe the hypothesis that "AsyncStream supports multiple concurrent consumption" is incorrect

@malhal you are correct that only a single 'consumer' will receive a particular value produced by the stream. perhaps i should have phrased this as 'concurrent iteration' rather than 'concurrent consumption' (or if you have a better suggestion, please share!). to clarify what i mean by this – `AsyncStream` supports behavior where a single backing 'source' of the stream can be iterated concurrently from two different concurrency domains. e.g. you can write this:

```swift
let stream = AsyncStream(...)
let t1 = Task.detached {
  for await _ in stream { print("task 1") }
}
let t2 = Task.detached {
  for await _ in stream { print("task 2") }
}
await (t1.value, t2.value)
```
and if you push a bunch of elements through the stream, they will be 'split' in some manner across the consuming `Task`s. however, `AsyncThrowingStream` _does not_ work this way – if at any point a second continuation is registered via an async iterator, the implementation raises a `fatalError()`. `AsyncStream` used to have the same behavior, but was changed at some point (see the linked forum posts for more info).

so the issue here is that the two types have some (IMO) confusing behavioral divergences. this is separate from the question of supporting 'multiple consumption' in the sense i believe you are thinking of it (every pending continuation receives every stream element).

**Comment by malhal** [comment]
I would suggest if a second consuming client is attempted that should fatal error, since after all it is misuse.

**Comment by malhal** [comment]
How about a compilation error? It is similar to trying to update a let.

**Comment by malhal** [comment]
> > I would suggest if a second consuming client is attempted then that should fatal error, since after all it is misuse.
> 
> i also suggested that [here](https://forums.swift.org/t/question-how-to-re-introduce-concurrent-iteration-runtime-error-for-asyncstream/71831), but there was concern about breaking client code.

I [replied](https://forums.swift.org/t/question-how-to-re-introduce-concurrent-iteration-runtime-error-for-asyncstream/71831/12) that it wouldn't break valid code which is likely the majority.

Perhaps another way of flagging it would be a Swift 6 style compilation error.

**Comment by malhal** [comment]
Any idea why they wanted AsyncStream to be Sendable? https://github.com/swiftlang/swift/pull/41713 Doesn't make sense to me since it should only be used once.

**Comment by NotTheNHK** [comment]
@jamieQ I think we can close this now. The PR is back on main again, and it doesn't seem like it will be reverted again (hopefully).

I was planning to update the documentation after all the reimplementation PRs have been merged, but we could go ahead and document the multi-consumer behavior now. Although I'm unsure how detailed it should be.

**Comment by NotTheNHK** [comment]
Ping @jamieQ.

**Comment by jamieQ** [comment]
@NotTheNHK If it's all the same to you, I'd prefer to leave this open until the docs are accurate (and the new implementation has "baked" for a bit longer). Personally the documentation being misleading was a major point of confusion for my past self, so it would be good to address that IMO. I appreciate your perseverance to improve the state of things here though!

**Comment by NotTheNHK** [comment]
@jamieQ That's fine with me. I'll prepare a PR to update this part of the documentation. 

I do still wonder how detailed it should be, though. Should we document the weakly defined FIFO behavior, or keep it more vague? @ktoso what do you think as well?
